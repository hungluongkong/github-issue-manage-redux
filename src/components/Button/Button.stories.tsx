import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';

import Button from '.';
import { PreviewArea, Code, DetailTable } from '../commonStyled';
import { ButtonType, ButtonSize } from '../../types';

// Detail of props
const propsDetail: JSX.Element = DetailTable({
  data: [
    {
      name: 'type',
      type: 'enum ButtonType { Default, Primary }',
      default: 'ButtonType.Primary',
    },
    {
      name: 'size',
      type: 'enum ButtonSize { Small, Normal }',
      default: 'ButtonSize.Normal',
    },
    { name: 'value', type: 'string', default: 'ButtonType.Primary' },
    {
      name: 'onClicked',
      type: 'function',
      description:
        '((event: React.MouseEvent<HTMLButtonElement, MouseEvent>) => void)',
    },
  ],
});

storiesOf('Button', module)
  .addDecorator((story) => <div style={{ margin: '0 30px' }}>{story()}</div>)
  .add('Button Style', () => (
    <>
      <h3>Button Types</h3>
      <PreviewArea>
        <Button
          type={ButtonType.Primary}
          onClick={action('click')}
          value="Primary Button"
        />
        <Button
          type={ButtonType.Default}
          onClick={action('click')}
          value="Default Button"
        />
      </PreviewArea>
      <Code>
        {`
          <Button type={ButtonType.Primary} onClick={action('click')}>Primary Button</Button>
          <Button type={ButtonType.Default} onClick={action('click')}>Default Button</Button>
        `}
      </Code>
      {propsDetail}
    </>
  ))
  .add('Button Size', () => (
    <>
      <h3>Button Size</h3>
      <PreviewArea>
        <Button
          onClick={action('click')}
          size={ButtonSize.Normal}
          value="Normal Button"
        />
        <Button
          onClick={action('click')}
          size={ButtonSize.Small}
          value="Small Button"
        />
      </PreviewArea>
      <Code>
        {`
          <Button onClick={action('click')} size={ButtonSize.Normal}>Normal Button</Button>
          <Button onClick={action('click')} size={ButtonSize.Small}>Small Button</Button>
        `}
      </Code>
      {propsDetail}
    </>
  ));
