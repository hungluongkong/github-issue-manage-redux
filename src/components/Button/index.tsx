import React from 'react';
import styled from 'styled-components';

import * as color from '../../theme/color';
import * as metrics from '../../theme/metrics';
import { ButtonType, ButtonProps, ButtonSize } from '../../types';

/**
 * Get css style for each type of button.
 * @param type type of button (enum)
 */
const typeToColor = (type: ButtonType): string =>
  type === ButtonType.Primary
    ? `
      background: ${color.PRIMARY_COLOR};
      color: ${color.DEFAULT_COLOR};
    `
    : `
      background: ${color.DEFAULT_COLOR};
      color: ${color.PRIMARY_COLOR};
    `;

/**
 * Styling button by buttonSize
 * @param bSize size of button (enum)
 */
const sizeToSize = (bSize: ButtonSize): string =>
  bSize === ButtonSize.Small
    ? `
      line-height: ${metrics.ButtonSizing.small.lineHeight};
      padding: 3px 5px;
      font-size: ${metrics.FontSize.Small};
    `
    : // Style normal button here
      '';

const StyledButton = styled.button<{ bType: ButtonType; size: ButtonSize }>`
  display: inline-block;
  min-width: 101px;
  text-align: center;
  font-size: ${metrics.FontSize.Default};
  line-height: 20px;
  padding: 6px 10px;
  vertical-align: middle;
  border-radius: 4px;
  border: ${color.PRIMARY_COLOR} 1px solid;

  :hover {
    opacity: ${metrics.OPACITY.hover};
    box-shadow: 0 5px 15px ${color.SHADOW_BORDER};
    cursor: pointer;
  }

  ${(props: { bType: ButtonType }): string => typeToColor(props.bType)}
  ${(props: { size: ButtonSize }): string => sizeToSize(props.size)}
`;

const Button = ({
  type = ButtonType.Primary,
  size = ButtonSize.Normal,
  value = '',
  onClick = () => {},
}: ButtonProps): JSX.Element => (
  <StyledButton
    {...{ size, onClick }}
    // bType: button type
    bType={type}
  >
    {value}
  </StyledButton>
);

export default Button;
