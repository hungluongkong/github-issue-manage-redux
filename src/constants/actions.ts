enum Action {
  GET_ISSUES = 'GET_ISSUES',
  GET_ISSUES_SUCCESS = 'GET_ISSUES_SUCCESS',
  GET_ISSUES_ERROR = 'GET_ISSUES_ERROR',

  UPDATE_ISSUE = 'UPDATE_ISSUE',
  TOGGLE_LOCK_ISSUE = 'TOGGLE_LOCK_ISSUE',
  UPDATE_ISSUE_SUCCESS = 'UPDATE_ISSUE_SUCCESS',

  REPORT_TASK = 'REPORT_TASK',

  TOGGLE_FORM = 'TOGGLE_FORM',
  TOGGLE_ISSUE = 'TOGGLE_ISSUE',
  TOGGLE_ERROR = 'TOGGLE_ERROR',
}

export default Action;
