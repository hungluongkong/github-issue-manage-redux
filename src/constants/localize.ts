
export const en = {
  FORM_TITLE: 'Title',
  FORM_CONTENT_TITLE: 'Content',
  FORM_TITLE_PLACEHOLDER: 'Your title here',
  FORM_CONTENT_PLACEHOLDER: 'Write something about this issue here',

  ISSUE_LIST_TITLE: 'Issue',
  ISSUE_LIST_NO_ISSUE: 'No Issue',

  ISSUE_CREATED_AT: 'Created at ',
  ISSUE_UPDATED_AT: 'Updated at ',

  BTN_ISSUE_LOCK: 'Lock Issue',
  BTN_ISSUE_UNLOCK: 'Unlock Issue',
  BTN_ISSUE_EDIT: 'Edit',
  BTN_ISSUE_ADD: 'New Issue',
  BTN_FORM_CLOSE: 'Close Form',
  BTN_SUBMIT: 'Submit',
  BTN_OK: 'Ok',
  BTN_CONFIRM: 'Confirm',

  WARN_EMPTY_TITLE: 'Title cannot be empty',

  INFORM_LOCK_DONE: 'Issue has locked!',
  INFORM_UNLOCK_DONE: 'Issue has unlocked!',
  INFORM_ADD_DONE: 'Issue added successful!',
  INFORM_UPDATE_DONE: 'Issue has updated!',
  INFORM_ISSUES_GET_DONE: 'Connect successful!',

  ERROR_SOMETHING_WRONG: 'Something went wrong!',
};
