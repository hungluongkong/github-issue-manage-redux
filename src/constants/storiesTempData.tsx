/**
 * Template response json string from github.
 * This is a list of issues.
 * URL: https://api.github.com/repos/hungluong-asnet/demo-api/issues
 */
export const issueJson: string = `
{
  "url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1",
  "repository_url": "https://api.github.com/repos/hungluong-asnet/demo-api",
  "labels_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1/labels{/name}",
  "comments_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1/comments",
  "events_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/1/events",
  "html_url": "https://github.com/hungluong-asnet/demo-api/issues/1",
  "id": 653740906,
  "node_id": "MDU6SXNzdWU2NTM3NDA5MDY=",
  "number": 1,
  "title": "isssue no 1",
  "user": {
    "login": "hungluong-asnet",
    "id": 68043715,
    "node_id": "MDQ6VXNlcjY4MDQzNzE1",
    "avatar_url": "https://avatars1.githubusercontent.com/u/68043715?v=4",
    "gravatar_id": "",
    "url": "https://api.github.com/users/hungluong-asnet",
    "html_url": "https://github.com/hungluong-asnet",
    "followers_url": "https://api.github.com/users/hungluong-asnet/followers",
    "following_url": "https://api.github.com/users/hungluong-asnet/following{/other_user}",
    "gists_url": "https://api.github.com/users/hungluong-asnet/gists{/gist_id}",
    "starred_url": "https://api.github.com/users/hungluong-asnet/starred{/owner}{/repo}",
    "subscriptions_url": "https://api.github.com/users/hungluong-asnet/subscriptions",
    "organizations_url": "https://api.github.com/users/hungluong-asnet/orgs",
    "repos_url": "https://api.github.com/users/hungluong-asnet/repos",
    "events_url": "https://api.github.com/users/hungluong-asnet/events{/privacy}",
    "received_events_url": "https://api.github.com/users/hungluong-asnet/received_events",
    "type": "User",
    "site_admin": false
  },
  "labels": [

  ],
  "state": "open",
  "locked": false,
  "assignee": null,
  "assignees": [

  ],
  "milestone": null,
  "comments": 0,
  "created_at": "2020-07-09T03:48:36Z",
  "updated_at": "2020-07-09T03:48:36Z",
  "closed_at": null,
  "author_association": "OWNER",
  "active_lock_reason": null,
  "body": "Nothing inside issue"
}
`;

export const issueListJson: string = `
  [
    {
      "url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2",
      "repository_url": "https://api.github.com/repos/hungluong-asnet/demo-api",
      "labels_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2/labels{/name}",
      "comments_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2/comments",
      "events_url": "https://api.github.com/repos/hungluong-asnet/demo-api/issues/2/events",
      "html_url": "https://github.com/hungluong-asnet/demo-api/issues/2",
      "id": 654622112,
      "node_id": "MDU6SXNzdWU2NTQ2MjIxMTI=",
      "number": 2,
      "title": "#practice #make #api",
      "user": {
        "login": "hungluong-asnet",
        "id": 68043715,
        "node_id": "MDQ6VXNlcjY4MDQzNzE1",
        "avatar_url": "https://avatars1.githubusercontent.com/u/68043715?v=4",
        "gravatar_id": "",
        "url": "https://api.github.com/users/hungluong-asnet",
        "html_url": "https://github.com/hungluong-asnet",
        "followers_url": "https://api.github.com/users/hungluong-asnet/followers",
        "following_url": "https://api.github.com/users/hungluong-asnet/following{/other_user}",
        "gists_url": "https://api.github.com/users/hungluong-asnet/gists{/gist_id}",
        "starred_url": "https://api.github.com/users/hungluong-asnet/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/hungluong-asnet/subscriptions",
        "organizations_url": "https://api.github.com/users/hungluong-asnet/orgs",
        "repos_url": "https://api.github.com/users/hungluong-asnet/repos",
        "events_url": "https://api.github.com/users/hungluong-asnet/events{/privacy}",
        "received_events_url": "https://api.github.com/users/hungluong-asnet/received_events",
        "type": "User",
        "site_admin": false
      },
      "labels": [

      ],
      "state": "open",
      "locked": false,
      "assignee": null,
      "assignees": [

      ],
      "milestone": null,
      "comments": 0,
      "created_at": "2020-07-10T08:47:16Z",
      "updated_at": "2020-07-10T08:47:16Z",
      "closed_at": null,
      "author_association": "OWNER",
      "active_lock_reason": null,
      "body": "**Don't know why this issue is here. and why we need to create it?**"
    },
    ${issueJson}
  ]`;
