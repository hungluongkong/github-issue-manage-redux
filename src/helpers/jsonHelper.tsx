/* eslint-disable camelcase */
import { Issue } from '../types';

export interface RawIssue {
  id: number;
  number: number;
  title: string;
  body: string;
  locked: boolean;
  url: string;
  created_at: string;
  updated_at: string;
}

export const RawToIssue = (raws: RawIssue[]): Issue[] =>
  raws
    .map((raw: RawIssue) => {
      const createdAt = new Date(raw.created_at);
      const updatedAt = new Date(raw.updated_at);

      // Destructor - remove unused properties
      const { id, number, url, title, body, locked } = raw;

      // Create Item
      return {
        id,
        number,
        title,
        body,
        locked,
        url,
        createdAt,
        updatedAt,
      } as Issue;
    })
    .sort((issue: Issue, nextIssue: Issue) => issue.number - nextIssue.number);

/**
 * Convert to list of issue from json
 * @param json response from github api
 */
export const JsonToIssueList = (json: string): Issue[] => {
  const raws: RawIssue[] = JSON.parse(json);

  return RawToIssue(raws);
};
