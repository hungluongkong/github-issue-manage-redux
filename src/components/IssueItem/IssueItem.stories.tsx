import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import IssueItem from '.';
import { JsonToIssueList } from '../../helpers/jsonHelper';
import { issueListJson } from '../../constants/storiesTempData';
import { PreviewArea, Code, DetailTable } from '../commonStyled';

// Detail of props
const propsDetail: JSX.Element = DetailTable({
  data: [
    { name: 'issue', type: 'interface Issue', description: 'Issue to display' },
  ],
});

const [issue] = JsonToIssueList(issueListJson);

storiesOf('Issue Item', module).add(
  'Issue',
  (): JSX.Element => (
    <>
      <h3>Default Item</h3>
      <PreviewArea>
        <IssueItem
          issue={issue}
          onToggleContent={action('toggle')}
          onToggleLock={action('toggle lock')}
        />
      </PreviewArea>
      <Code>
        {`
          <IssueItem issue={issue} onToggleContent={action('toggle')} onToggleLock={action('toggle lock')} />
        `}
      </Code>
      {propsDetail}
    </>
  ),
);
