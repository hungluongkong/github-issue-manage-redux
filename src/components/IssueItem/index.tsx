import React from 'react';
import styled from 'styled-components';
import Button from '../Button';

import * as color from '../../theme/color';
import * as metrics from '../../theme/metrics';
import * as localize from '../../constants/localize';
import { isEqualAllProps } from '../../helpers/propsHelper';
import {
  IssueProps,
  ButtonType,
} from '../../types';

// Main container
const IssueWrapper = styled.div<{isLocked: boolean}>`
  display: flex;
  box-sizing: border-box;
  padding: 5px;
  justify-content: space-between;
  background: ${(props) => (props.isLocked ? color.ISSUE_BG_LOCKED : color.ISSUE_BG_DEFAULT)};

  :hover {
    opacity: ${metrics.OPACITY.hover};
    box-shadow: 2px 1px 2px 1px ${color.SHADOW_BORDER};
  }

  :before {
    content: ${(props) => (props.isLocked ? '"ðŸ”’"' : '"ðŸ”“"')};
    padding: 7px 0 10px 5px;
    font-size: ${metrics.FontSize.Icon};
  }
`;

// Display Id in github of issue. On left - middle
const IssueID = styled.div`
  font-weight: 600;
  text-align: center;
  padding: 10px 20px 10px 0px;
  min-width: 30px;
`;

// Wrap title & date. Export for test
export const TitleWrapper = styled.div`
  flex-grow: 1;
  :hover {
    cursor: pointer;
  }
`;

// Big issue title
export const IssueTitle = styled.a`
  font-weight: 700;
  display: block;
`;

// Edited/create date under the title.
const EditTime = styled.span`
  display: block;
  color: ${color.SMALL_TEXT_COLOR};
  font-size: ${metrics.FontSize.Small};
  margin-top: 4px;
  line-height: 1.25;
`;

const IssueItem = ({
  issue,
  onToggleContent,
  onToggleLock,
}: IssueProps): JSX.Element => (
  <IssueWrapper isLocked={issue.locked}>
    {/* Issue number is id of issue in user's storage list */}
    <IssueID>#{issue.number}</IssueID>

    {/* Body of issue item */}
    <TitleWrapper onClick={() => { onToggleContent(issue); }}>
      <IssueTitle>{issue.title}</IssueTitle>
      <EditTime>
        {
          // Compare edit/create timestamp
          new Date(issue.createdAt).getTime() === new Date(issue.updatedAt).getTime()
            ? localize.en.ISSUE_CREATED_AT
            : localize.en.ISSUE_UPDATED_AT
        }
        { new Date(issue.updatedAt).toLocaleDateString() }
      </EditTime>
    </TitleWrapper>
    {
      <Button
        type={issue.locked ? ButtonType.Primary : ButtonType.Default}
        onClick={() => { onToggleLock(issue); }}
        value={issue.locked ? localize.en.BTN_ISSUE_UNLOCK : localize.en.BTN_ISSUE_LOCK}
      />
    }
  </IssueWrapper>
);

export default React.memo(IssueItem, isEqualAllProps);
