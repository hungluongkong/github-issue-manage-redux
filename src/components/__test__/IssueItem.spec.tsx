import React from 'react';
import { shallow } from 'enzyme';
import { create } from 'react-test-renderer';

import Button from '../Button';
import IssueItem, {
  TitleWrapper,
  IssueTitle,
} from '../IssueItem';

import { JsonToIssueList } from '../../helpers/jsonHelper';
import { issueListJson } from '../../constants/storiesTempData';

describe('Unit test for IssueItem component', () => {
  const [issue] = JsonToIssueList(issueListJson);

  it('snapshot', () => {
    const issueItem = create(
      <IssueItem
        issue={issue}
        onToggleContent={jest.fn()}
        onToggleLock={jest.fn()}
      />,
    );
    expect(issueItem).toMatchSnapshot();
  });

  it('simulate toggle issue content', () => {
    const onToggleContent = jest.fn();
    const issueItem = shallow(
      <IssueItem
        issue={issue}
        onToggleContent={onToggleContent}
        onToggleLock={() => {}}
      />,
    );

    issueItem.find(TitleWrapper).simulate('click');
    expect(onToggleContent).toBeCalledTimes(1);
  });

  it('simulate edit button click', () => {
    const onToggleLock = jest.fn();
    const issueItem = shallow(
      <IssueItem
        issue={issue}
        onToggleContent={() => {}}
        onToggleLock={onToggleLock}
      />,
    );

    issueItem.find(Button).simulate('click');
    expect(onToggleLock).toBeCalledTimes(1);
  });

  it('should display issue title', () => {
    const issueItem = shallow(
      <IssueItem
        issue={issue}
        onToggleContent={() => {}}
        onToggleLock={() => {}}
      />,
    );

    expect(issueItem.find(IssueTitle).text()).toEqual(issue.title);
  });
});
