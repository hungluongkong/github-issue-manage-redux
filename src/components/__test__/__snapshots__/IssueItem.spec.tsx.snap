// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit test for IssueItem component snapshot 1`] = `
<div
  className="sc-AxirZ bggFi"
>
  <div
    className="sc-AxiKw cEQHGP"
  >
    #
    1
  </div>
  <div
    className="sc-AxhCb bREgZd"
    onClick={[Function]}
  >
    <a
      className="sc-AxhUy ogOxz"
    >
      isssue no 1
    </a>
    <span
      className="sc-AxgMl gaiwNa"
    >
      Created at 
      7/9/2020
    </span>
  </div>
  <button
    className="sc-AxjAm gZMYpY"
    onClick={[Function]}
    size={1}
  >
    Lock Issue
  </button>
</div>
`;
